File: nomad/src/net/sf/nmedit/nomad/main/task/PatchLoader.java
/* Copyright (C) 2006 Christian Schneider
 * 
 * This file is part of Nomad.
 * 
 * Nomad is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Nomad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Nomad; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * Created on Jan 23, 2006
 */
package net.sf.nmedit.nomad.main.task;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;

import net.sf.nmedit.jpatch.clavia.nordmodular.v3_03.Patch;
import net.sf.nmedit.jpatch.io.FileSource;
import net.sf.nmedit.jpatch.io.PatchDecoder;
import net.sf.nmedit.jpatch.spi.PatchDecoderProvider;
import net.sf.nmedit.jpatch.spi.PatchImplementation;
import net.sf.nmedit.nomad.main.Nomad;
import net.sf.nmedit.nomad.patch.ui.PatchDocument;
import net.sf.nmedit.nomad.patch.ui.PatchUI;

public class PatchLoader {

	private final Nomad nomad;
	//private ArrayList<PatchFile> fileList = new ArrayList<PatchFile>();
	
	public PatchLoader(Nomad nomad) {
		this.nomad = nomad; }
	public void loadPatch(String file) {
		loadPatch(new String[]{file});
	}

	public void loadPatch(File[] files) {
		String[] sfiles = new String[files.length];
		for (int i=files.length-1;i>=0;i--)
			sfiles[i]=files[i].getAbsolutePath();
		
		loadPatch(sfiles);
	}
    
    public static String extractPatchName(String fileName)
    {
        String patchName = fileName.substring(0,fileName.lastIndexOf(".pch"));
        int sep = patchName.lastIndexOf('/');
        if (sep<0) sep = patchName.lastIndexOf('\\');
        if (sep>=0)
            patchName = patchName.substring(sep+1);
        
        return patchName;
    }

    public void loadPatch(String[] files) {
        PatchImplementation impl = nomad.getPatchImplementation();
        PatchDecoderProvider provider;
        try
        {
            provider = impl.getPatchDecoderProvider(FileSource.class);
        }
        catch (Exception e)
        {
            e.printStackTrace();
            return ;
        }
        
        PatchDocument first = null;
        
        for (String name:files)
        {
            try 
            {
                PatchDecoder decoder = provider.createDecoder(impl);
                
                FileReader fr = new FileReader(name);
                decoder.decode( new FileSource(fr) );
                
                Patch patch = (Patch) decoder.getPatch();
                
                PatchUI ui = PatchUI.newInstance(patch);
                PatchDocument doc = new PatchDocument(ui);
                if (first==null) first = doc;
                doc.setFile(new File(name));
                String pName = extractPatchName(name);

                patch.setName(pName);
                patch.getHistory().clear();
                patch.getHistory().setModified(false);
                
                synchronized (nomad)
                {
                    nomad.addPatchDocument(doc);
                }
                
                fr.close();
                
            } catch (FileNotFoundException e) 
            {
                e.printStackTrace();
            } 
            catch (Throwable t) 
            {
                System.out.println("Error while loading patch.");
                t.printStackTrace();
            }
        }
        
        if (first!=null)
        {
            nomad.getDocumentContainer().setSelection(first);
        }
        
    }
    
}

File: archive/patchloader/src/patchloader.cc
/*
    Nord Modular patch loader
    Copyright (C) 2003 Marcus Andersson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <unistd.h>

#include "nmprotocol/mididriver.h"
#include "nmprotocol/midiexception.h"
#include "nmprotocol/patchmessage.h"
#include "nmprotocol/nmprotocol.h"

#include "pdl/pdlexception.h"

#include "ppf/ppfexception.h"

#include "nmpatch/patch.h"
#include "nmpatch/modulesection.h"
#include "nmpatch/patchexception.h"

void load(string, string, int, string, string, string);

extern char *optarg;
extern int optind;

int main(int argc, char** argv)
{
  printf("Patchloader version %s, Copyright (C) 2003 Marcus Andersson\n"
	 "Patchloader comes with ABSOLUTELY NO WARRANTY. This is free "
	 "software,\nand you are welcome to redistribute it under certain "
	 "conditions.\n", VERSION);

  try {
    string patchname, drivername, input, output;
    bool error = false;
    int slot = 0;

    int option;
    while ((option = getopt(argc, argv, "n:s:d:i:o:")) != -1) {

      switch(option) {

      case 'n':
	patchname = string(optarg);
	break;
	
      case 'd':
	drivername = string(optarg);
	break;

      case 'i':
	input = string(optarg);
	break;
	
      case 'o':
	output = string(optarg);
	break;
	
      case 's':
	sscanf(optarg, "%d", &slot);
	if (slot < 0 || slot > 3) {
	  error = true;
	}
	break;

      case '?':
	error = true;

      default:
	break;
      }
    }
    
    if (error || optind >= argc) {
      printf("usage: %s \\\n"
	     "        -n patchname -s {0,1,2,3} -d mididriver \\\n"
	     "        -i midiinput -o midioutput patchfile\n\n",
	     argv[0]);

      printf("Available midi drivers:\n");
      MidiDriver::StringList drivers = MidiDriver::getDrivers();
      for (MidiDriver::StringList::iterator i = drivers.begin();
	   i != drivers.end(); i++) {
	printf(" %s\n", (*i).c_str());
	MidiDriver* driver = MidiDriver::createDriver(*i);
	printf("  inputs:");
	MidiDriver::StringList inputs = driver->getMidiInputPorts();
	for (MidiDriver::StringList::iterator j = inputs.begin();
	     j != inputs.end(); j++) {
	  printf(" %s", (*j).c_str());
	}
	printf("\n  outputs:");
	MidiDriver::StringList outputs = driver->getMidiOutputPorts();
	for (MidiDriver::StringList::iterator j = outputs.begin();
	     j != outputs.end(); j++) {
	  printf(" %s", (*j).c_str());
	}
	printf("\n\n");
      }
      exit(1);
    }

    string filename = argv[optind];

    if (patchname == "") {
      int start = filename.find_last_of("/\\", filename.size()) + 1;
      int end = filename.find_first_of(".", start);
      patchname = filename.substr(start, end-start);
    }

    printf("Loading patch...\n");
    load(filename, patchname, slot, drivername, input, output);
    printf("Done.\n");

  }
  catch (MidiException& exception) {
    printf("MidiException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
  catch (PDLException& exception) {
    printf("PDLException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
  catch (PatchException& exception) {
    printf("PatchException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
  catch (ppf::PPFException& exception) {
    printf("PPFException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
}

void load(string filename, string patchname, int slot,
	  string drivername, string input, string output)
{
  Patch* patch = new Patch(filename);
  patch->setName(patchname);
  Patch::PositionList sectionEndPositions;
  BitStream patchStream = patch->serialize(&sectionEndPositions);
  PatchMessage patchMessage(patchStream, sectionEndPositions);
  patchMessage.setSlot(slot);
  
  MidiDriver* driver = MidiDriver::createDriver(drivername);
  driver->connect(input, output);
  
  NMProtocol nmProtocol(driver);
  
  nmProtocol.send(&patchMessage);
  
  while(!nmProtocol.sendQueueIsEmpty()) {
    nmProtocol.heartbeat();
    usleep(10000);
  }
  
  driver->disconnect();
}
